core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala

```diff
@@ -151,9 +151,9 @@ private final class TransactionalProducerStageLogic[K, V, P](stage: Transactiona
   private def maybeCommitTransaction(beginNewTransaction: Boolean = true): Unit = {
     val awaitingConf = awaitingConfirmation.get
     batchOffsets match {
-      case batch: NonemptyTransactionBatch if awaitingConf == 0 =>
+      case batch: NonemptyTransactionBatch if awaitingConf == 0 && !hasBeenPulled(stage.in) =>
```

Comment:
Can you explain why we need this logic?

Reply:
With this PR, the transaction commit is split into two synchronous blocks in the Producer stage. That means that in between, a message could be sent to the stage which would be sent to the Kafka producer, but the transaction has not yet been started.

So this guard was needed to be sure that no message will come in between the the committing of transaction and beginning of the new one.

Reply:
Still didn't quite get it. I was hoping that suspending demand when transaction is in progress will be enough.
Also, I'm wondering if we just shouldn't block waiting for internal commit.
`kafkaProducer.commitTrasnaction()` is already blocking the whole stage waiting on a network call. And `internalCommit` is just waiting for local memory change, so it shouldn't add much...
That would simplify the code a bit.

Reply:
Making `internalCommit`just send a message would also be possible. Seems like `TransactionalProducerStage` does not have to wait for the SourceActor to acknowledge getting the message, right?

Reply:
> I was hoping that suspending demand when transaction is in progress will be enough.

Calling `suspendDemand` does make sure that `stage.in` is not going to be pulled anymore from that point on, but it could have been pulled from before.

> Seems like TransactionalProducerStage does not have to wait for the SourceActor to acknowledge getting the message, right?

Yes, currently it looks like that's the case. I'll need to think about it a bit more.